modRlp/src/org/aion/rlp/CompactEncoder.java

```diff
@@ -112,16 +110,35 @@
 
         if (nibbles[nibbles.length - 1] == TERMINATOR) {
             terminator = 1;
-            nibbles = copyOf(nibbles, nibbles.length - 1);
         }
-        int oddlen = nibbles.length % 2;
+        int oddlen = (nibbles.length - terminator) % 2;
         int flag = 2 * terminator + oddlen;
         if (oddlen != 0) {
-            byte[] flags = new byte[]{(byte) flag};
-            nibbles = concatenate(flags, nibbles);
+            byte[] ret;
+            if (terminator == 0) {
+                ret = new byte[nibbles.length + 1];
+                ret[0] = (byte) flag;
+                System.arraycopy(nibbles, 0, ret, 1, nibbles.length);
+            } else { // terminator == 1
+                ret = new byte[nibbles.length];
+                ret[0] = (byte) flag;
+                System.arraycopy(nibbles, 0, ret, 1, nibbles.length - 1);
+            }
+            nibbles = ret;
         } else {
-            byte[] flags = new byte[]{(byte) flag, 0};
-            nibbles = concatenate(flags, nibbles);
+            byte[] ret;
+            if (terminator == 0) {
+                ret = new byte[nibbles.length + 2];
+                ret[0] = (byte) flag;
+                ret[1] = 0;
+                System.arraycopy(nibbles, 0, ret, 2, nibbles.length);
+            } else { // terminator == 1
+                ret = new byte[nibbles.length + 1];
+                ret[0] = (byte) flag;
+                ret[1] = 0;
+                System.arraycopy(nibbles, 0, ret, 2, nibbles.length - 1);
+            }
+            nibbles = ret;
```

Comment:
This method seems like can be refactored cause the nibbles been copied twice during the logic (including the following ByteArrayOutputStream write). Considering find a way to avoid too many copies in this method.

Reply:
optimized in PR #572 commit 50ce43b

Reply:
good job!